syntaxes/src/expression/grammar.ts

```diff
@@ -0,0 +1,886 @@
+import { GrammarDefinition } from '../types';
+
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+export const expression: GrammarDefinition = {
+  scopeName: 'expression.ng',
+  injectionSelector: 'L:text.html -comment',
+  patterns: [{ include: '#ngExpression' }],
+  repository: {
+    ngExpression: {
+      name: 'meta.expression.ng',
+      patterns: [
+        {
+          include: '#tsString',
+        },
+        {
+          include: '#literal',
+        },
+        {
+          include: '#ternaryExpression',
+        },
+        {
+          include: '#expressionOperators',
+        },
+        {
+          include: '#functionCall',
+        },
+        {
+          include: '#identifiers',
+        },
+        {
+          include: '#parenExpression',
+        },
+        {
+          include: '#punctuationComma',
+        },
+        {
+          include: '#punctuationAccessor',
+        },
+      ],
+    },
+
+    arrayLiteral: {
+      name: 'meta.array.literal.ts',
+      begin: /\[/,
+      beginCaptures: {
+        '0': {
+          name: 'meta.brace.square.ts',
+        },
+      },
+      end: /\]/,
+      endCaptures: {
+        '0': {
+          name: 'meta.brace.square.ts',
+        },
+      },
+      patterns: [
+        {
+          include: '#ngExpression',
+        },
+        {
+          include: '#punctuationComma',
+        },
+      ],
+    },
+
+    booleanLiteral: {
+      patterns: [
+        {
+          name: 'constant.language.boolean.true.ts',
+          match: /(?<!\.|\$)\btrue\b(?!\$)/,
+        },
+        {
+          name: 'constant.language.boolean.false.ts',
+          match: /(?<!\.|\$)\bfalse\b(?!\$)/,
+        },
+      ],
+    },
+
+    expressionOperators: {
+      patterns: [
+        {
+          match: /((?<!\|)(\|)(?!\|))\s?([a-zA-Z0-9\-\_]*)/,
```

Comment:
I am confused by what this is matching - a `|` lookbehind, then a  `|`, then another `|` via a lookahead? Can you provide an example where this is the case? Please correct me if I am misinterpreting the pattern.

Reply:
This is used to match pipes.

I am matching a `|` that doesn't have a `|` before and another after, and are followed by a alphanumeric characters.

I made it this way to prevent it from interacting with the `||` operator, as it is allowed in the same level as the pipes.

Reply:
This seems like a pretty specific match. I think we could instead demarcate statements by pipe and semicolon, and then make the `||` operator a level below as something allowed as an expression in the statement. What do you think?

Reply:
The problem with demarcating statements by semicolon in this case is that it is not mandatory, and often are removed automatically by code formatters. 

The only way I found to make pipes work properly was inserting in at the first level of the expression pattern alongside all of the operators.

It is now pretty solid and it won't conflict with the `||` operator because it is first in the pattern array, so it gets the top priority. And also the matching itself excludes each other.

I won't object in finding another way to do it, but IMO, I don't think it is that urgent.

Reply:
A potential simplification would be to replace the negative lookahead with a `{1}` and replace the `*` in the second group with `+` (since you can't have a pipe with no identifier) — but you do still need the negative lookbehind so it's honestly not a huge improvement.

One thing I would recommend is removing the extra parens around the `\|` in the middle of the first group, because that creates a duplicate capture group (compare the Details tab [here](https://regexr.com/4tk7o) and [here](https://regexr.com/4tk7r) with the match selected)

Reply:
@dannymcgee this is not resolved yet, but It's fine creating a separate comment